### PR TITLE
Feature/remove unused kernel model fields

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
As was mentioned in #929 we don't need host specific fields in kernel. Another issue that we addressing here, since #1263 not all fields on device have addrspace, which results in broken `opencl:gpu` test cases. This PR is all about removing `parent`, `meminfo` and `sycl_queue` from device representation of usm arrays. 

What was done:
- Use proper datamodel in few places
- Remove `parent`, `meminfo` and `sycl_queue` fields from the data model used by the `USMNdArray` type in kernels
- Introduce kernel specific allocation of array for local/private memories and views
- Organize datamodel specific tests

Checklist:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?

Closes: https://github.com/IntelPython/numba-dpex/issues/929

### CODE REVIEW

File: numba_dpex/core/kernel_interface/arrayobj.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# SPDX-FileCopyrightText: 2012 - 2024 Anaconda Inc.
+# SPDX-FileCopyrightText: 2024 Intel Corporation
+#
+# SPDX-License-Identifier: BSD-2-Clause
+
+"""
+This package contains implementation of some numpy.np.arrayobj functions without
+parent and meminfo fields required, because they don't make sense on device.
+These functions intended to be used only in kernel targets.
```
Comment (ZzEeKkAa) [inline numba_dpex/core/kernel_interface/arrayobj.py:9]: Add use cases

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 2]
```diff
@@ -1078,14 +1082,29 @@ def getitem_arraynd_intp(context, builder, sig, args):
     that when returning a view of a dpnp.ndarray the sycl::queue pointer
     member in the LLVM IR struct gets properly updated.
     """
-    ret = np_getitem_arraynd_intp(context, builder, sig, args)
+    is_kernel = isinstance(context, DpexKernelTargetContext)
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:1085]: A better name is needed. How about `getitem_call_in_kernel`?

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 3]
```diff
@@ -1078,14 +1082,29 @@ def getitem_arraynd_intp(context, builder, sig, args):
...
+    ary, idx = args
+
+    assert aryty.ndim >= 1
+    ary = make_array(aryty)(context, builder, ary)
 
-    if isinstance(sig.return_type, DpnpNdArray):
+    res = _getitem_array_generic(
+        context, builder, sig.return_type, aryty, ary, (idxty,), (idx,)
+    )
+    ret = impl_ret_borrowed(context, builder, sig.return_type, res)
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:1100]: What does `impl_ret_borrowed` do?

**Review by diptorupd — COMMENTED** [review]
Few minor things, but overall good to go. Thank you!

File: numba_dpex/dpnp_iface/arrayobj.py (hunk 3 above)
Comment (ZzEeKkAa) [reply numba_dpex/dpnp_iface/arrayobj.py:1100]: I don't know it is copy-paste from dpex

File: numba_dpex/dpnp_iface/arrayobj.py (hunk 2 above)
Comment (ZzEeKkAa) [reply numba_dpex/dpnp_iface/arrayobj.py:1085]: Whatever sounds better here. I don't have preferences. This function is really short. Ideally it won't be an issue with target specific overload.

File: numba_dpex/core/utils/kernel_launcher.py [hunk 4]
```diff
@@ -265,82 +266,54 @@ def _build_array_arg(  # pylint: disable=too-many-arguments
...
-            arg_num=arg_num,
-        )
-        arg_num += 1
         # Argument nitems
         self._build_array_attr_arg(
-            array_val=array_val,
-            array_attr_pos=array_data_model.get_field_position("nitems"),
-            array_attr_ty=array_data_model.get_member_fe_type("nitems"),
+            array_val=host_array_val,
+            array_attr_pos=host_array_data_model.get_field_position("nitems"),
```
[the remark this answers is not in the document]
Comment (ZzEeKkAa) [reply numba_dpex/core/utils/kernel_launcher.py:277]: It needs proper refactoring. Did my best to give reader some clue

**Review by diptorupd — APPROVED** [review]

**Pull request merged**